### DISCUSSION TITLE
How can I parse "NaN", "-Infinity", "Infinity" doubles generated by Jackson?

### DISCUSSION THREAD

**Discussion by mbasmanova** [nlohmann/json#2443] [Q&A] [answered]
#### What is the issue you have?

I'm getting `[json.exception.type_error.302] type must be number, but is string` error when parsing NaN double values generated by Jackson: "totalSize":"NaN". That JSON also have positive and negative double infinities:

```
            "lowValue":"-Infinity",
            "highValue":"Infinity",
```

Is it possible to parse these?

#### Which compiler and operating system are you using?

$ gcc -v
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 11.0.0 (clang-1100.0.33.17)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin

**Comment by nlohmann** [comment, marked as answer]
This is not possible, because these values are not supported by JSON.

**Comment by mbasmanova** [comment]
@nlohmann Niels, thanks for clarifying.

**Comment by HarryDC** [comment]
This has been brought up multiple times now, as much as appreciate the work being done here, with everything else is there no way to come up with a reasonable way to handle NANs?  I'm guessing here that in javascript i can assign `null` to a member, so the standard works perfectly well there. in C++ we can't, what is so wrong about either creation an extension point or adopting what other implementations are doing ?

**Reply by SylvainFogale** [reply]
I registered here for my first and only post of all times, as I had just now the same situation when interacting with matlab, and C# serialization. 

Anecdote : **I fixed it for my own use in "the worst possible way"** (be warned - I did not read all the discussions, not all the documentation, I am not fluent with the grammar topics, etc ...) : the "one line hack" simply consisted of inserting the case as follow, when a string is received where a numerical value is expected, a NaN numerical value is generated.
void from_json(const BasicJsonType& j, ArithmeticType& val)
{
[...]
        case value_t::string:
        {
            val = std::nan( j.template get_ptr<const typename BasicJsonType::string_t*>()->c_str() );
            break;
        }
[...] other fall-through cases that throw
}

If at least there would be a way of handling the exception in the middle of the parsing of a complex class and pursue rather than throw everything with some values filled, some not.
I also want to stress that the IEEE754 ensured that double sqrt(double x) outputs a double when x>=0 and when x<0 aswell, but the standard for JSON cannot handle the latter case. It imply more or less to create subnode at every leaf of the tree with a double replaced by an "optionnal variant" that could be there or not there, could be a double numerical value, could be not.

Specification : I suppose the approach shall be more refined as follow, but this may be beyond my ability :
=> firstly it must be a user defined option to not enforce the standard, as was done for the "comments", exactly because that is against the standard.
=> secondly it shall apply only when the expected type is float or double, there is no concept of NaN for integer types (although in this case one practice it that NaN converts to the largest negative number the integer type can hold)
=>It would have to take into account the syntax "value" : "NaN" where the right hand is a string type (has quotes). If the expected type is float or double, the received type is a string, and that the string belong to a list of authorized NaN and infinity keywords, it could be implemented.
=>Ideally, it would have to take into account the syntax "value" : NaN or "value" : 1#NAN (the language basic uses 1#INF 1#NAN) where the right hand is not a string (not with quotes) but not starting with a digit. I would expect this to be handled by the parser who shall still conclude the right hand is a value of type double rather than a string, as much as it handle already (I hope) the scientific notation in all its splendor 1e6 1.0e6 .1e-6 -.1E-6 by allowing some form of escaping for some character patterns.

References : I do not want to mess up the existing discussions, it is of course important to refer to, and understand, the following :
https://github.com/nlohmann/json/issues/3494 -> won't fix
https://json.nlohmann.me/features/comments/ -> The Harmful Consequences of the Robustness Principle
https://fhur.me/posts/theres-no-such-thing-as-a-one-line-change

Conclusion : Beyond the quick fix my hope is that more competent programmers may mature the idea and feature of practical interest : this topic is not recurring because we users want to fuck the standard, it is because the practical cases are recurring.

Last comment nevertheless, this lib is wonderful, works hand in hand with https://app.quicktype.io/ : thousands of line of C++ code generated from a single JSON example, and works perfectly, that you to the author and contributors for all the work for developing and maintaining this tool.
